### PR TITLE
Added additional reaction controls to channel_id and message structs

### PR DESCRIPTION
**Pull request by Collin-Swish**
#2532 In reference to this issue.

### CODE REVIEW

File: src/model/channel/channel_id.rs [hunk 1]
```diff
@@ -279,6 +279,22 @@ impl ChannelId {
             )
             .await
     }
+    /// Deletes all of the [`Reaction`]s associated with the provided message id.
+    /// 
+    /// **Note**: Requires the [Manage Messages] permission.
+    ///
+    /// [Manage Messages]: Permissions::MANAGE_MESSAGES
```
Comment (mkrasnitski) [inline src/model/channel/channel_id.rs]: Should include an `# Errors` section.

File: src/model/channel/message.rs [hunk 2]
```diff
@@ -264,6 +264,36 @@ impl Message {
...
+    /// # Errors
+    /// 
+    /// Returns [`Error::Http`] if the currend user did not perform the reaction,
+    /// and lacks permission
+    /// 
+    /// [Manage Messages]: Permissions::MANAGE_MESSAGES
+    #[inline]
+    pub async fn delete_reaction(
+        self,
+        cache_http: impl CacheHttp,
```
Comment (mkrasnitski) [inline src/model/channel/message.rs]: If a permissions check isn't done on the `CacheHttp`, better to change this to `http: impl AsRef<Http>`.

File: src/model/channel/message.rs [hunk 3]
```diff
@@ -264,6 +264,36 @@ impl Message {
...
+        cache_http
+            .http()
+            .as_ref()
+            .delete_reaction(
+                self.channel_id.0,
+                self.id.0,
+                user_id.map(|uid| uid.0),
+                &reaction_type.into(),
+            )
+            .await
```
Comment (mkrasnitski) [inline src/model/channel/message.rs]: This should be reworked to just a wrapper around `ChannelId::delete_reaction`. The same goes for `Message::delete_reactions`, now that `ChannelId::delete_reactions` exists.

File: src/model/channel/message.rs [hunk 4]
```diff
@@ -264,6 +264,36 @@ impl Message {
         cache_http.http().as_ref().delete_message_reactions(self.channel_id.0, self.id.0).await
     }
 
+    /// Remove reaction from the message given an [`Emoji`] or unicode character
+    /// 
+    /// **Note**: Requires the [Manage Messages] permission, _if_ the current
+    /// user did not perform the reaction
```
Comment (mkrasnitski) [inline src/model/channel/message.rs]:
```suggestion
    /// **Note**: Requires the [Manage Messages] permission, _if_ the current user did not perform
    /// the reaction.
```
formatting

File: src/model/channel/message.rs [hunk 5]
```diff
@@ -264,6 +264,36 @@ impl Message {
...
+    /// 
+    /// # Errors
+    /// 
+    /// Returns [`Error::Http`] if the currend user did not perform the reaction,
+    /// and lacks permission
```
Comment (mkrasnitski) [inline src/model/channel/message.rs]:
```suggestion
    /// Returns [`Error::Http`] if the current user did not perform the reaction, or lacks
    /// permission.
```
formatting

**Review by mkrasnitski — CHANGES_REQUESTED** [review]
Could you add a `GuildChannel::delete_reactions` wrapper method as well? (A `PrivateChannel::delete_reactions` method doesn't make sense, because only reactions by the current user can be removed in a private channel.)

Also, make sure to run `cargo +nightly fmt`.

File: src/model/channel/message.rs (hunk 2 above)
Comment (Collin-Swish) [reply src/model/channel/message.rs]: Good point. The change has been made.

File: src/model/channel/guild_channel.rs [hunk 6]
```diff
@@ -439,6 +439,26 @@ impl GuildChannel {
...
+    /// Returns [`Error::Http`] if the current user lacks permission
+    ///
+    /// [Manage Messages]: Permissions::MANAGE_MESSAGES
+    #[inline]
+    pub async fn delete_reactions(
+        &self,
+        http: impl AsRef<Http>,
+        message_id: impl Into<MessageId>,
+    ) -> Result<()> {
+        http.as_ref().delete_message_reactions(self.id.0, message_id.into().0).await
```
Comment (mkrasnitski) [inline src/model/channel/guild_channel.rs]:
```suggestion
        self.id.delete_reactions(http, message_id).await
```

File: src/model/channel/message.rs [hunk 7]
```diff
@@ -264,6 +264,32 @@ impl Message {
         cache_http.http().as_ref().delete_message_reactions(self.channel_id.0, self.id.0).await
     }
 
+    /// Remove reaction from the message given an [`Emoji`] or unicode character
```
Comment (mkrasnitski) [inline src/model/channel/message.rs]:
```suggestion
    /// Deletes the given [`Reaction`] from the message.
```
Adjusted wording to mirror docs for `ChannelId::delete_reaction`.

**Review by mkrasnitski — CHANGES_REQUESTED** [review]
Looks pretty good. Could you also change the implementation of `Message::delete_reactions` to be:
```rust
self.channel_id.delete_reactions(cache_http.http(), self.id).await
```

**Review by mkrasnitski — APPROVED** [review]

**Review by arqunis — APPROVED** [review]

**Comment by mkrasnitski** [comment]
This slipped past in review, but `Message::delete_reaction` should take `&self` instead of `self` as it doesn't make sense to take ownership. Conversely, `ChannelId::delete_reactions` should in fact take `self` instead of `&self`, because `ChannelId` is `Copy`.

**Pull request merged**